database/gormdb_submission.go

```diff
@@ -65,13 +66,20 @@ func (db *GormDB) CreateSubmission(submission *pb.Submission) error {
 	// TODO(meling) temporary transformation of submission data
 	transform(submission)
 
-	if submission.BuildInfo != nil {
-		if err := db.conn.Save(submission.BuildInfo).Error; err != nil {
+	// A submission will have a build info without any ID, using save will create a duplicate build info
+	// in case the submission already exists in the database. We have to update build info explicitly.
+	// There should be a less hacky way to do it.
```

Comment:
Not sure how to do it less hacky. But is there a test that exposes this problem? If not, we should add one. I added print statements inside the if-block below, and no prints where found during testing. So suspect there isn't any tests yet...

I think the reason could be that I may have removed the `gorm:foreignKey` tag in `score.proto`:

```proto
    uint64 SubmissionID = 2 [(go.field) = {tags: 'gorm:"foreignKey:ID"'}];
```

I used to have this, but I have been hacking on this stuff, and I might have removed it without any of the tests breaking, and thereby thinking it would be okay...

Note also, in `web/submissions_test.go:L433-434,441-442` we actually set the `ID` and `SubmissionID`. These should not be set when calling `CreateSubmission`, I think... All tests still passes, when I remove these.

Reply:
It is something I've noticed about the gorm's Save() behavior: if we save a struct that has another struct (or a slice of these) as its field (like `Submission.BuildInfo`), Save will properly create records for the nested struct(s) if there are no primary IDs in those, but just ignores them (neither creating nor updating) when there are primary IDs. Doesn't look like it has something with foreign keys to do.

Added a new test for this behavior.